### PR TITLE
Update PV annotation according to volume scheduled condition

### PR DESCRIPTION
**Pull request by derekbit**
Currently, the Condition["Schedule"].message is still empty even if there is a scheduling failure. We can update it with a verbose error message which is a combination of the errors, e.g. insufficient storage or insufficient storage, disk not found. Then, the verbose error message is updated to PV annotation.

https://github.com/longhorn/longhorn/issues/3529

### CODE REVIEW

File: controller/volume_controller.go
Comment (innobead) [inline controller/volume_controller.go:56]: We should use longhorn namespace instead of customizing for any downstream project.

Comment (derekbit) [reply controller/volume_controller.go:56]: Agree.

File: types/types.go
Comment (shuo-wu) [inline types/types.go:120]: NIT: will this be more straightforward? `PVAnnotationLonghornVolumeSchedulingError = "longhorn.io/volume-scheduling-error"`

File: controller/volume_controller.go
Comment (shuo-wu) [inline controller/volume_controller.go:1060]: I would recommend aggregating and setting the scheduling failure message outside of the `for` loop.
The logic may be a little bit tricky. VolumeController will mark the volume scheduling condition as False **as long as** there is **one** replica failing at it. But the error message is the **summarization** of **all** scheduling failures and should be set after the iteration.

File: controller/volume_controller.go
Comment (derekbit) [reply controller/volume_controller.go:1043]: Yes. But what's the purpose of keeping the replica name? Would you like to show the error of each replica in the aggregated error message? e.g.
"summary of the error message: replica1: <error message>, replica2: <error message>...."

Comment (shuo-wu) [reply controller/volume_controller.go:1043]: Do you think it's necessary? It seems that most of the replica scheduling failure reasons would be similar/identical.

Comment (derekbit) [reply controller/volume_controller.go:1043]: I don't think we need the replica names and lengthy and similar/identical failure reasons.

Comment (shuo-wu) [reply controller/volume_controller.go:1043]: OK, then ignore the related review comments

File: scheduler/replica_scheduler.go
Comment (innobead) [inline scheduler/replica_scheduler.go:100]: It seems the returned verbose error type `map[string]struct{}` frequently used here, so I would suggest having a type alias which is better for debugging/development.

**Review by innobead — APPROVED** [review]
In general, LGTM. After tackling Shuo's comments, we can merge this.

Need to backport to 1.2.x, so we need to have the error types in v1beta API when doing backport. Thank you!

File: controller/volume_controller.go
Comment (innobead) [reply controller/volume_controller.go:1060]: Yeah, indeed this is tricky, but I feel the intermediate messages should be fine but leave it to @derekbit to decide.

File: scheduler/replica_scheduler.go
Comment (derekbit) [reply scheduler/replica_scheduler.go:100]: okay

File: scheduler/replica_scheduler.go
Comment (shuo-wu) [inline scheduler/replica_scheduler.go:64]: Create and use a function here `NewVerboseErrors(strList...)`

File: scheduler/replica_scheduler.go
Comment (shuo-wu) [inline scheduler/replica_scheduler.go:100]: Always return `util.VerboseErrors` rather than a string

**Review by shuo-wu — APPROVED** [review]
In general LGTM

File: manager/kubernetes.go
Comment (innobead) [inline manager/kubernetes.go:70]: I would prefer this should be reconciled in the controller only, so we don't need to have sperate update logic in more than one place. The downside will be some delay for the update, but it will be finally consistent. WDYT?

File: k8s/pkg/apis/longhorn/v1beta2/node.go
Comment (innobead) [inline k8s/pkg/apis/longhorn/v1beta2/node.go:44]: Change to alias with the spec property -> `SchedulerErrorHardNodeAffinityNotSatisfied`, the same as the description needed to adjust.

File: scheduler/replica_scheduler.go
Comment (innobead) [inline scheduler/replica_scheduler.go:102]: Can we use string formatting w/ interpolation to add `schedulingReplica.Spec.HardNodeAffinity` into the error message for having much detailed info? like fmt.Sprintf(longhorn.SchedulerErrorHardNodeAffinityNotSatisfied, schedulingReplica.Spec.HardNodeAffinity).

File: scheduler/replica_scheduler.go
Comment (innobead) [inline scheduler/replica_scheduler.go:284]: The same, if possible, add a formatted error w/ node name, so it's better to know which node has no disks available? 

The same as below errors.

File: scheduler/replica_scheduler.go
Comment (innobead) [inline scheduler/replica_scheduler.go:304]: Should we have a formatted message for the disk name or uuid?

File: scheduler/replica_scheduler.go
Comment (innobead) [inline scheduler/replica_scheduler.go:311]: Suggest changing the error much meaningful like `longhorn.SchedulerErrorSchedulingSettingsGetError`

**Review by innobead — CHANGES_REQUESTED** [review]

File: k8s/pkg/apis/longhorn/v1beta2/node.go
Comment (derekbit) [reply k8s/pkg/apis/longhorn/v1beta2/node.go:39]: Hard to know the error. This error is caused by the failure while getting setting from etcd.

Comment (innobead) [reply k8s/pkg/apis/longhorn/v1beta2/node.go:39]: I mean, we need to change a meaningful error like "scheduling settings failed to retrieve".

File: controller/volume_controller.go
Comment (innobead) [inline controller/volume_controller.go:1116]: This is good. We should just warn out instead of returning the error, because this is an optional update outside Longhorn resources.

File: scheduler/replica_scheduler.go
Comment (derekbit) [reply scheduler/replica_scheduler.go:284]: If there are tens or hundreds of disks, the annotation become very lengthy. IMO, the final result is that the replica can be scheduled in each disk and on each node. Current annotation value gives enough clues and can helps the administrator to understand what happen in the system.
For more information, user can the details in longhorn logs.

Comment (innobead) [reply scheduler/replica_scheduler.go:284]: Agreed that the concern is fair. Let's keep it as is, but just ensure there will be no duplicated errors in the aggregated error.

File: util/util.go
Comment (derekbit) [reply util/util.go:899]: After checking  https://github.com/hashicorp/go-multierror and https://github.com/uber-go/multierr, both are not good for our use case. They can append and combine the errors and store errors in a slice struct, but they cannot de-duplicate the same errors. We still need to implement our `combine` function using `errors.Is`, but the bigO is n^2, looks not a good idea...

Comment (innobead) [reply util/util.go:899]: Then we just keep the current way as it first.

File: controller/volume_controller.go
Comment (derekbit) [reply controller/volume_controller.go:1108]: Yes, I think we need `GetCondition` for getting the `Reason`. Two failure reasons in the scheduled condition:  `VolumeConditionReasonLocalReplicaSchedulingFailure` and `VolumeConditionReasonReplicaSchedulingFailure`.

File: controller/volume_controller.go
Comment (derekbit) [reply controller/volume_controller.go:1117]: Good idea, thank you.

File: manager/kubernetes.go
Comment (derekbit) [reply manager/kubernetes.go:70]: Yes. If the delay is acceptable, I prefer your idea.

File: util/multierror.go
Comment (innobead) [inline util/multierror.go:9]: Usually, we will use Error interface for error, but here use string instead. This is fine for now for the specific need. However, it will be good to improve this later, to make the MultiError support Error args instead of a raw string because you can use err.Error() to get the error message.

**Review by innobead — APPROVED** [review]
LGTM

File: util/multierror.go
Comment (innobead) [reply util/multierror.go:9]: Also, I would prefer using error.go instead of a specific multierror.go, because it will be much extensible in the future, but this is quite minor now :slightly_smiling_face:

Comment (derekbit) [reply util/multierror.go:9]: Yes, the naming is a bit of weird. In order to do a fast deduplication of errors, `map[string]error` is preferred, but I'm not quiet sure `ERRORREPLICASCHEDULEXXX` should string or error for now...
Maybe we should write our own `multierror` package following errors interface for our use case and replace current impl with it. Let me think about it more.

Comment (innobead) [reply util/multierror.go:9]: yeah, I know you want to use map for set purpose.

**Pull request merged**
